File: src/main/scala/com/emofy/controllers/LoginFormController.scala
package com.emofy.controllers

import com.emofy.services.UserService
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.{GetMapping, PostMapping, RequestBody, ResponseBody, ModelAttribute}
import com.emofy.models.User
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.{HttpStatus, ResponseEntity}

@Controller
class LoginFormController @Autowired()(userService: UserService) {

  @GetMapping(Array("/login"))
  def showLoginForm(): String = "loginForm"

  @PostMapping(Array("/login"))
  @ResponseBody
  def loginUser(@ModelAttribute user: User): ResponseEntity[String] = {


    
    try {


      //user can be authenticated either by email or username
      val paramProvided: String = Option(user.username).getOrElse(user.email)


      userService.loginUser(paramProvided, user.password) match {
        case Some(user) =>
            // User authentication successful, you can access the user object here
            ResponseEntity.status(HttpStatus.CREATED).body("Hi, " + user.username + "!. Welcome back!")
            
        case None =>
            // User authentication failed
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body("Incorrect credentials") 
      }
      ResponseEntity.status(HttpStatus.CREATED).body("User "+ user.username + " logged successfully!")
    } catch {
      case e: Exception =>
        // Handle failure
        ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("Failed to login user")
    }
  }

}

File: login/src/main/java/com/example/demo/controller/LoginFormController.java
package com.example.demo.controller;

import com.example.demo.models.User;
import com.example.demo.services.UserService;
import com.example.demo.services.JwtTokenService;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.*;
import java.util.Optional;
import jakarta.servlet.http.HttpServletResponse;

@Controller
public class LoginFormController {

    private final UserService userService;

    @Autowired
    public LoginFormController(UserService userService) {
        this.userService = userService;

    }

    @GetMapping("/login")
    public String showLoginForm() {
        return "loginForm";
    }

    @PostMapping("/login")
    @ResponseBody
    public ResponseEntity<String> loginUser(@ModelAttribute User user, HttpServletResponse response) {
        try {
            
            // User can be authenticated either by email or username
       
            
            String paramProvided = user.getUsername() != null ? user.getUsername() : user.getEmail();
    
            Optional<User> authenticatedUser = userService.loginUser(paramProvided, user.getPassword(), response);
            if (authenticatedUser.isPresent()) {
                // User authentication successful
                response.sendRedirect("http://emofy-image-storage:8081"/*"http://localhost:8081" */);
                return ResponseEntity.status(HttpStatus.CREATED).body("Hi, " + authenticatedUser.get().getUsername() + ". Welcome back!");
            } else {
                // User authentication failed
                return ResponseEntity.status(HttpStatus.BAD_REQUEST).body("Incorrect credentials");
            }
        } catch (Exception e) {
            // Handle failure
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("Failed to login user");
        }
    }
}
